### PR TITLE
Enhance the logic for the merge selector

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduced a special (experimental) mode of a merge selector for MergeTree tables which makes it more aggressive for the partitions that are close to the limit by the number of parts. It is controlled by the `merge_selector_use_blurry_base` MergeTree-level setting.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 2d76f8b1d6e2d550d6525ee1e23d3ba2bb261689 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70645/2d76f8b1d6e2d550d6525ee1e23d3ba2bb261689/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/ad7c74c549343391b5320bcf95dc2b44b4f04d40/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeSettings.cpp [hunk 1]
```diff
@@ -52,6 +52,10 @@ namespace ErrorCodes
     M(Bool, load_existing_rows_count_for_old_parts, false, "Whether to load existing_rows_count for existing parts. If false, existing_rows_count will be equal to rows_count for existing parts.", 0) \
     M(Bool, use_compact_variant_discriminators_serialization, true, "Use compact version of Variant discriminators serialization.", 0) \
     \
+    /** Merge selector settings. */ \
+    M(Bool, merge_selector_use_blurry_base, false, "Special mode for merge selector which makes it more aggressive when the table reaches the threshold for the number of parts.", 0) \
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeSettings.cpp:56]: Let's make it separate selector after  https://github.com/ClickHouse/ClickHouse/pull/70647

**Review by alesapin — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp [hunk 2]
```diff
@@ -528,18 +531,28 @@ SelectPartsDecision MergeTreeDataMergerMutator::selectPartsToMergeFromRanges(
...
         auto merge_selector_algorithm = (*data_settings)[MergeTreeSetting::merge_selector_algorithm];
+
         std::any merge_settings;
-        if (merge_selector_algorithm == MergeSelectorAlgorithm::SIMPLE)
+        if (merge_selector_algorithm == MergeSelectorAlgorithm::SIMPLE
+            || merge_selector_algorithm == MergeSelectorAlgorithm::STOCHASTIC_SIMPLE)
         {
             SimpleMergeSelector::Settings simple_merge_settings;
             /// Override value from table settings
+            simple_merge_settings.window_size = (*data_settings)[MergeTreeSetting::merge_selector_window_size];
```
Comment (nikitamikhaylov) [inline src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp:541]: `parts_to_throw_insert` should be also initialized here.

File: src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp [hunk 3]
```diff
@@ -528,18 +531,29 @@ SelectPartsDecision MergeTreeDataMergerMutator::selectPartsToMergeFromRanges(
...
             if (!(*data_settings)[MergeTreeSetting::min_age_to_force_merge_on_partition_only])
                 simple_merge_settings.min_age_to_force_merge = (*data_settings)[MergeTreeSetting::min_age_to_force_merge_seconds];
 
             if (aggressive)
                 simple_merge_settings.base = 1;
 
+            if (merge_selector_algorithm == MergeSelectorAlgorithm::STOCHASTIC_SIMPLE)
+            {
+                simple_merge_settings.parts_to_throw_insert = (*data_settings)[MergeTreeSetting::parts_to_throw_insert];
+                simple_merge_settings.use_blurry_base = true;
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp:552]:
```suggestion
                simple_merge_settings.use_blurry_base = false;
```

**Comment by alesapin** [comment]
Failures unrelated and I need it.

**Pull request merged**